lib/consul/index.js

```diff
@@ -72,27 +82,77 @@ var render_obj = function(parts, prefix, obj) {
 
     if (_.isObject(val)) return render_obj(parts, prefix + '/' + encodeURIComponent(key), val)
 
-    parts.push({'key': prefix + '/' + encodeURIComponent(key), 'value': val});
+    // Value of all KV objects should be treated as strings
+    parts.push({'key': prefix + '/' + encodeURIComponent(key), 'value': val.toString()});
   });
 }
 
 /**
- * Walk through obj, creating a Consul write operation for each kv pair.  Objects are treated
- * as parents of subtrees.  Arrays are ignored since they add annoying problems (multiple array
- * entries can have the same value, so what do we do then?).
+ * After computing the diff between the object and the existing KV records,
+ * perform the necessary consul operations such that the state in consul matches
+ * the diff.
  */
-var populate_kvs_from_object = function(branch, prefix, obj, cb) {
+var populate_kvs_from_object = function(branch, prefix, obj, existing_kvs, cb) {
+  var write_kvs = [];
+  var delete_kvs = [];
+  var candidate_kvs = [];
 
-  var writes = [];
+  render_obj(candidate_kvs, prefix, obj);
 
-  render_obj(writes, prefix, obj);
+  // This avoids unnecessary copying if there are no existing KV records.
+  if (existing_kvs.length > 0) {
+    diff_kvs(write_kvs, delete_kvs, candidate_kvs, existing_kvs);
+  } else {
+    write_kvs = candidate_kvs;
+  }
 
-  logger.debug('Expandable file %s yielded %s keys', prefix, writes.length);
+  logger.debug('Expandable file %s yielded %s keys (%s writes, %s deletes)',
+               prefix,
+               candidate_kvs.length,
+               write_kvs.length,
+               delete_kvs.length);
 
-  cb = _.after(writes.length, cb);
+  // This helps ensure that writes happen after all deletes
+  // have been processed.
+  var do_writes = _.after(delete_kvs.length, function(err) {
+    if (err) return cb(err);
 
-  writes.forEach(function(write) {
-    write_content_to_consul(write.key, write.value, cb);
+    if (write_kvs.length > 0) {
+      cb = _.after(write_kvs.length, cb);
+      write_kvs.forEach(function(write) {
+        write_content_to_consul(write.key, write.value, cb);
+      });
+    } else {
+      cb();
+    }
+  });
+
+  if (delete_kvs.length > 0) {
+    delete_kvs.forEach(function(del) {
+      delete_from_consul(del.key, do_writes);
+    });
+  } else {
+    do_writes();
+  }
+};
+
+/**
+ * Compare the list of candidate KV records to the list of existing KV records.
+ * Based on the comparison, populate the provided lists for write and delete
+ * operations against the KV store.
+ */
+var diff_kvs = function(write_kvs, delete_kvs, candidate_kvs, existing_kvs) {
```

Comment:
The two `for` loops in here is the only thing that could be concerning in terms of performance. If the diff (candidate_kvs) or the existing KV tree (existing_kvs) is big enough, this call could take some time.